### PR TITLE
Product Gallery > Pager block: Add support to interactivity API

### LINKED ISSUE #10630: Product Gallery Block: Replace the image in the Large Image Block with the clicked page from the Pager block
## Description
The Pager block is one of the inner blocks contained within the Product Gallery block. When a shopper clicks on a page from the Pager block, the Large Image block should replace its image with the image that matches the clicked page.

## Outcomes

- The Large Image block should display the image that matches the clicked page from the Pager block.

## Acceptance criteria
- Upon clicking a page from the block, the Large Image block should promptly reflect image that matches the clicked page.
- If a shopper clicks on a page that matches the image already displayed in the Large Image block, no action should occur.

### PR DESCRIPTION
**Pull request by thealexandrelara**
## What
This PR adds support to the Interactivity API on the Paget block, allowing shoppers to navigate between the gallery images without having to reload the page.

Fixes #10630

## Why

This is part of our effort to implement the new Product Gallery block.

## Testing Instructions

_Please consider any edge cases this change may have, and also other areas of the product this may impact._

1. From your WordPress dashboard, go to Appearance > Themes. Make sure you have a block-based theme installed and activated. If not, you can install one from the Add New option. Block-based themes include "Twenty-twenty Two," "Twenty-twenty Three," etc.
2. On the left-hand side menu, click on Appearance > Editor > Templates
3. Find and select the 'Single Product' template from the list.
4. When the Classic Product Template renders, click on Transform into Blocks. This will transform the Classic template in a block template if you haven't done it before.
5. In the block library, search for the 'Product Gallery' block. Click on it to add the block to the template.
6. Check that the Pager block is displayed on the Editor, and its display mode is set to "Dots":
CleanShot 2023-09-12 at 19 06 04
7. Click on save and visit a product page that contains multiple gallery images, make sure the block is correctly being displayed and when selecting a different page, the Large Image changes to reflect the selected image.
8. Change the display mode to "Digits" (follow step 6) and repeat step 7.
9. Change the display mode to "Off" (follow step 6), click on Save and visit a product page that contains multiple gallery images, make sure the block is not displayed.

* [x] Do not include in the Testing Notes 
* [ ] Should be tested by the development team exclusively 

## Screenshots or screencast

| Display mode: Dots | Display mode: Digits | Display mode: Off |
| ------ | ----- | ----- |
| CleanShot 2023-09-12 at 19 11 44 | CleanShot 2023-09-12 at 19 11 15 | CleanShot 2023-09-12 at 19 12 34 |

## WooCommerce Visibility

Required:

* [ ] WooCommerce Core
* [ ] Feature plugin
* [x] Experimental
* [ ] N/A

### CODE REVIEW

File: src/BlockTypes/ProductGalleryPager.php [hunk 1]
```diff
@@ -37,96 +39,128 @@ protected function get_block_type_uses_context() {
...
+
+			if ( $p->next_tag() ) {
+				$p->set_attribute(
+					'data-wc-context',
+					wp_json_encode(
+						array(
+							'isSelected'  => $is_first_pager_item,
+							'woocommerce' => array( 'imageId' => strval( $product_gallery_image_id ) ),
+						)
+					)
```
Comment (gigitux) [inline src/BlockTypes/ProductGalleryPager.php]: Everything should be inside under the woocommerce key. Also, I am not sure that `isSelected` is necessary

File: src/BlockTypes/ProductGalleryPager.php [hunk 2]
```diff
@@ -37,96 +39,128 @@ protected function get_block_type_uses_context() {
...
-			</svg>'
-		);
+	protected function get_block_type_script( $key = null ) {
+		$script = [
+			'handle'       => 'wc-' . $this->block_name . '-frontend',
+			'path'         => $this->asset_api->get_block_asset_build_path( $this->block_name . '-frontend' ),
+			'dependencies' => [ 'wc-interactivity' ],
+		];
+
+		return $key ? $script[ $key ] : $script;
```
Comment (gigitux) [inline src/BlockTypes/ProductGalleryPager.php]: Is it necessary?

File: src/BlockTypes/ProductGalleryPager.php [hunk 3]
```diff
@@ -37,96 +39,128 @@ protected function get_block_type_uses_context() {
...
+		$html = '';
+
+		foreach ( $product_gallery_images_ids as $key => $product_gallery_image_id ) {
+			if ( $key >= $number_of_thumbnails ) {
+				break;
+			}
+
+			$is_first_pager_item = 0 === $key;
+			$pager_item          = sprintf(
+				'<li class="wp-block-woocommerce-product-gallery__pager-item %2$s">%1$s</li>',
```
Comment (gigitux) [inline src/BlockTypes/ProductGalleryPager.php]: For custom classes, let's use `wc-block-`.

File: src/BlockTypes/ProductGalleryPager.php [hunk 4]
```diff
@@ -37,96 +39,128 @@ protected function get_block_type_uses_context() {
...
+				'dots' === $pager_display_mode ? $this->get_dot_icon( $is_first_pager_item ) : $key + 1,
+				$is_first_pager_item ? 'is-active' : ''
```
Comment (gigitux) [inline src/BlockTypes/ProductGalleryPager.php]: `is-active` is too much generic. Let's use a class with `wc` prefix. https://github.com/woocommerce/woocommerce-blocks/blob/73151afd5929b747fb8bd98c37337e22a830ee2e/docs/contributors/coding-guidelines.md/#L15-L18

**Review by gigitux — CHANGES_REQUESTED** [review]
The PR works great, but there are some things to address. Also, I think that we should add E2E tests to catch any regression in the future.

File: src/BlockTypes/ProductGalleryPager.php (hunk 4 above)
Comment (thealexandrelara) [reply src/BlockTypes/ProductGalleryPager.php]: I've tried to change the `is-active` to `.wc-block-woocommerce-product-gallery__pager-item--is-active`; however, it breaks Interactivity API when using with the `data-wc-class` directive, for example:
```
$p->set_attribute(
	'data-wc-class--wc-block-woocommerce-product-gallery__pager-item--is-active',
	'selectors.woocommerce.isSelected'
);
```

It seems that this directive does not accept the `__` or `--` combination. The only way I've managed to make it work was to change the classname to `wc-block-woocommerce-product-gallery-pager-item-is-active` but it goes against the guidelines that you shared. Any suggestions?

File: src/BlockTypes/ProductGalleryPager.php (hunk 1 above)
Comment (thealexandrelara) [reply src/BlockTypes/ProductGalleryPager.php]: you're right, I forgot to remove it from there when I refactored the block 🤦

File: src/BlockTypes/ProductGalleryPager.php (hunk 2 above)
Comment (thealexandrelara) [reply src/BlockTypes/ProductGalleryPager.php]: I believe it is not necessary anymore, it is remaining from a previous version of this block.

File: src/BlockTypes/ProductGalleryPager.php (hunk 4 above)
Comment (gigitux) [reply src/BlockTypes/ProductGalleryPager.php]: Good point! I would update with `wc-block-woocommerce-product-gallery-pager-item-is-active`and adding a comment that explains why. What do you think?

**Review by gigitux — APPROVED** [review]
Thanks! I’m pre-approving this PR, given that the class name that you suggested is reasonable. Also, you are working on E2E tests in another PR: https://github.com/woocommerce/woocommerce-blocks/pull/11001! 💪

Great work!

File: src/BlockTypes/ProductGalleryPager.php (hunk 4 above)
Comment (thealexandrelara) [reply src/BlockTypes/ProductGalleryPager.php]: I followed your suggestion to use `wc-block-woocommerce-product-gallery-pager-item-is-active` and left a comment there explaining why

Comment (luisherranz) [reply src/BlockTypes/ProductGalleryPager.php]: We fixed that in Core in this PR, but we haven't updated Woo's codebase yet:

- https://github.com/WordPress/gutenberg/pull/53337

Feel free to replace the old reg exp with the new one yourselves 🙂

Comment (thealexandrelara) [reply src/BlockTypes/ProductGalleryPager.php]: Hey @luisherranz,
Thank you for sharing the solution for this with us! I just opened a PR that implements this: https://github.com/woocommerce/woocommerce-blocks/pull/11034

**Pull request merged**